File: MeiqiaSDK-Push-Signature-Example/signature.js
// ECMAScript 5
// the method to read raw body in Express
// app.configure(function() {
// 	//. . .
// 	app.use(express.bodyParser());
// 	app.use(function(req, res, next) {
// 	        var data = '';
// 	        req.setEncoding('utf8');
// 	        req.on('data', function(chunk) { 
// 	            data += chunk;
// 	        });
// 	        req.on('end', function() {
// 	            req.rawBody = data;
// 	        });
// 	        next();
// 	    });
	
// 	//. . .
// });
// 
// Example:
// if(DTsiner(req.rawBody) != req.getHeaders('Authorization')){
//
// }
var crypto = require('crypto')

function DTSigner(raw_body, key){
	var hash_str = crypto.createHmac('sha1', key).update(raw_body).digest('hex');
	var sign_str = new Buffer(hash_str).toString('base64')
	.replace(/\+/g, '-')
    .replace(/\//g, '_');
	return 'meiqia_sign:' + sign_str;
}

File: MeiqiaSDK-Push-Signature-Example/signature.rb
require 'base64'
require 'openssl'

# In rack apps that include Rails and Sinatra,
# raw_data is request.raw_body which only can be read once.

# Example:
# error!("Authorization is invalid", 401) unless MeiqiaToolsSigner.get(request.raw_body, secrect_key) == headers["Authorization"]

class MeiqiaToolsSigner
  def self.get(raw_body, key)
    hash_str = OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('sha1'), key, raw_body)
    sign_str = Base64.urlsafe_encode64(hash_str)
    "meiqia_sign:#{sign_str}"
  end
end
